File: FirstApp/Kitty.vb
Imports System.Data.OleDb
Imports FirstApp.Utility

Public Class Kitty
    Implements IEquatable(Of Kitty)

#Region "Properties"
    Public ReadOnly Property KittyUID As Integer = -1

    Public Property KittyNo As Integer = -1

    Public Property KittyType As Integer = -1

    Public Property KittyInterest As Integer = -1

    Public Property Duration As Integer = -1

    Public Property CustomerID As Integer = -1

    Public Property Notes As String = ""

    Public Property IsAvailed As Boolean = False

    Public Property IsInactive As Boolean = False

    Public Property Record As New Dictionary(Of Date, Integer)

    Public Property TransactionsRecord As New Dictionary(Of Date, String)

    Private Property Initialized As Boolean = False

#End Region

    Sub New(Optional UID As Integer = -1, Optional InitializeKitty As Boolean = False, Optional completely As Boolean = False)
        If UID = -1 Then Exit Sub
        KittyUID = UID
        If InitializeKitty Then Initialize(completely)
    End Sub

    Sub Initialize(Optional InitializeCompletely As Boolean = True)
        If KittyUID = -1 Then Exit Sub
        If Initialized Then Exit Sub
        Try
            Dim dr As OleDbDataReader = DataReader("Select * From Kitty_Data Where KittyUID=" & KittyUID)
            While dr.Read
                KittyNo = dr("KittyNo")
                CustomerID = dr("CoustID")
                KittyType = dr("KittyType")
                KittyInterest = dr("Interest")
                Duration = dr("Duration")
                IsAvailed = If(dr("Availed") = "False", False, True)
                IsInactive = If(dr("InActive") = "False", False, True)
                Try : Notes = dr("Notes").ToString.Replace(apostropheReplacer, "'") : Catch : End Try
                If InitializeCompletely Then
                    AddRecord(dr("Dates"))
                    Initialized = True
                End If
            End While
        Catch ex As Exception
            MessageBox.Show("[Kitty]/Initialize Error: " + ex.Message)
        End Try
    End Sub

    Sub Save()
        If KittyUID = -1 Then
            If CustomerID = -1 Then
                MessageBox.Show("[Kitty] Attempt To Save A Kitty Without A Customer To Own It. Can't Proceed Further Without A CustomerId.", "Illegal Selection", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Exit Sub
            End If
            If MessageBox.Show("[Kitty] Do You Want To Add New Kitty ?", "Confirm", MessageBoxButtons.YesNo, MessageBoxIcon.Question) = DialogResult.No Then Exit Sub
            SqlCommand(
                "INSERT INTO Kitty_Data (CoustID,KittyNo,KittyType,Dates,Interest,Duration,Availed,InActive,Notes,TransactionDetails)" &
                $"VALUES({CustomerID},{KittyNo},{KittyType},'{RecordString()}',{KittyInterest},{Duration},'{IsAvailed}','{IsInactive}','{Notes.Replace("'", apostropheReplacer)}', '{TransactionDetailsString()}')")
        Else
            If CustomerID = -1 Then Exit Sub
            If MessageBox.Show("[Kitty] Do You Want To Update This Kitty ?", "Confirm", MessageBoxButtons.YesNo, MessageBoxIcon.Question) = DialogResult.No Then Exit Sub
            SqlCommand(
                $"UPDATE Kitty_Data set CoustID={CustomerID},KittyNo={KittyNo},KittyType={KittyType},Dates='{RecordString()}',Interest={KittyInterest},Duration={Duration}" &
                $",Availed ='{IsAvailed}',InActive='{IsInactive}',Notes='{Notes.Replace("'", apostropheReplacer)}',TransactionDetails='{TransactionDetailsString()}' WHERE KittyUID={KittyUID}")
        End If
    End Sub

    Public Sub Dissolve()
        Try
            If MessageBox.Show("Are You Sure That You Want To Dissolve This Kitty ?", "Confirm", MessageBoxButtons.YesNo, MessageBoxIcon.Question) = DialogResult.No Then Exit Sub
            SqlCommand($"DELETE FROM Kitty_Data WHERE KittyUID={KittyUID}")
        Catch ex As Exception
            MessageBox.Show($"Error [Kitty]/Dissolve: {ex.Message}")
        End Try
    End Sub

    Public ReadOnly Property Owner As Customer
        Get
            Try
                Return New Customer(CustomerID)
            Catch ex As Exception
                MessageBox.Show("Error [Kitty]/Owner This Kitty DoesNot Have A Associated Customer With It.", "Illegeal Selection", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Return Nothing
            End Try
        End Get
    End Property

    Sub InitializeTransactions()
        Dim Transactions As String = ""
        Try
            Dim dr As OleDbDataReader = DataReader("Select TransactionDetails From Kitty_Data Where KittyUID=" & KittyUID)
            While dr.Read
                Try : Transactions = dr("TransactionDetails") : Catch ex As Exception : End Try
            End While
        Catch ex As Exception
            MessageBox.Show("[Kitty]/Initialize Transactions Error: " + ex.Message)
        End Try
        If Transactions IsNot Nothing AndAlso Transactions.Length > 1 Then
            AddTransactionDetails(Transactions)
        End If
    End Sub

    Sub AddTransactionDetails(entry As String)
        Try
            For Each _transaction In Split(entry, "|,|", , CompareMethod.Text)
                If _transaction.Length < 11 Then Continue For
                Dim _date = _transaction.Substring(0, 10).Trim
                Dim _details As String = _transaction.Substring(10)
                TransactionsRecord.Item(_date) = _details
            Next
        Catch ex As Exception
            MessageBox.Show("[Kitty]/AddTransactionDetails Error: " + ex.Message)
        End Try
    End Sub

    Function TransactionDetailsString(Optional Seperator As String = "|,|")
        Try
            If TransactionsRecord Is Nothing Then Return ""
            For Each i In TransactionsRecord.Values
                If i.Length > 1 Then
                    Return String.Join(Seperator, TransactionsRecord.Select(Function(Kvp) $"{Kvp.Key.ToShortDateString}{Kvp.Value}").ToArray)
                End If
            Next
            Return ""
        Catch ex As Exception
            MessageBox.Show("Can't Get TransactionsRecord As List [Kitty]: " + ex.Message)
            Return ""
        End Try
    End Function

    Public Function AddRecord(_entry As String) As Boolean
        Try
            If String.IsNullOrEmpty(_entry) Then Return False
            For Each i As String In _entry.Split(",")
                Dim _entryDate = i.Substring(0, 10).Trim
                Dim _entryAmount As Integer = i.Substring(10)
                If Record.ContainsKey(_entryDate) Then
                    If Record.Item(_entryDate) + _entryAmount < KittyType Then
                        Record.Remove(_entryDate)
                        If TransactionsRecord.ContainsKey(_entryDate) Then TransactionsRecord.Remove(_entryDate)
                    Else
                        Record.Item(_entryDate) += _entryAmount
                    End If
                Else
                    If _entryAmount >= KittyType Then
                        Record.Add(_entryDate, _entryAmount)
                        If Not TransactionsRecord.ContainsKey(_entryDate) Then
                            TransactionsRecord.Add(_entryDate, "")
                        End If
                    End If
                End If
            Next

            Dim sorted = From pair In Record
                         Order By pair.Key

            Record = sorted.ToDictionary(Function(k) k.Key, Function(v) v.Value)
            Return True
        Catch ex As Exception
            MessageBox.Show("Can't Add Entry To Record [Kitty]: " + ex.Message)
            Return False
        End Try
    End Function

    Public Function RecordString(Optional Seperator As String = ",") As String
        Try
            If Record Is Nothing Then Return Nothing
            Return String.Join(Seperator, Record.Select(Function(Kvp) $"{Kvp.Key.ToShortDateString}{Kvp.Value}").ToArray)
        Catch ex As Exception
            MessageBox.Show("Can't Get Record As List [Kitty]: " + ex.Message)
            Return Nothing
        End Try
    End Function

    Public Function TotalAmount() As Integer
        Try
            Return KittyType * Duration
        Catch ex As Exception
            MessageBox.Show("Can't Get Total Amount [Kitty]: " + ex.Message)
            Return Nothing
        End Try
    End Function

    Public Function GivenAmount() As Integer
        Try
            Dim total As Integer = Record.Values.Sum()
            Return total
        Catch ex As Exception
            MessageBox.Show("Can't Get Given Amount [Kitty]: " + ex.Message)
            Return Nothing
        End Try
    End Function

    Public Function GetInstalmentsCompleted() As Integer
        Try
            Return Int(GivenAmount() / KittyType())
        Catch ex As Exception
            MessageBox.Show("Can't Get Completed Instalments Completed [Kitty]: " + ex.Message)
            Return Nothing
        End Try
    End Function

    Public Function GetInstalmentsPending() As Integer  'Instalments Pending Till Date
        Try
            If Record.Count = 0 Then Return Nothing
            Dim FirstDate As Date = Record.First.Key

            Dim ExpectedDate As Date = FirstDate
            Dim ExpectedInstalments As Integer = 0

            While ExpectedDate < Today  ' Adding One To Get Instalments Expected Until Today (It Will Add One Extra Month At Last)
                ExpectedDate = ExpectedDate.AddMonths(1)
                ExpectedInstalments += 1
            End While

            ExpectedInstalments -= 1 ' Subtracting One To Get Instalments Expected Until Today

            ExpectedInstalments -= GetInstalmentsCompleted() - 1 ' Subtracting Instalments Given And One More Instalment That Is The First Instalment

            If ExpectedInstalments > GetInstalments_LeftForMaturity() Then ' If Instalments Due Past The Date Of Maturity
                ExpectedInstalments = GetInstalments_LeftForMaturity()
            End If

            If GetInstalments_LeftForMaturity() = 0 Then Return -1 ' If Matured -> Return -1

            If ExpectedInstalments < 0 Then 'If SomeBody Paid More Than Necessary
                Return 0
            Else
                Return ExpectedInstalments
            End If
        Catch ex As Exception
            MessageBox.Show("Can't Get Pending Instalments [Kitty]: " + ex.Message)
            Return Nothing
        End Try
    End Function

    Public Function GetInstalments_LeftForMaturity() As Integer 'Instalments Pending For Maturity
        Try
            Return Int(Duration - GetInstalmentsCompleted())
        Catch ex As Exception
            MessageBox.Show("Can't Get Instalments Pending For Maturity [Kitty]: " + ex.Message)
            Return Nothing
        End Try
    End Function

    Public Function IsMatured() As Boolean
        Try
            If GivenAmount() >= KittyType * Duration Then
                Return True
            Else
                Return False
            End If
        Catch ex As Exception
            MessageBox.Show($"Error Occured While Checking Maturity Of Kitty [Kitty]: {ex.Message}")
            Return False
        End Try
    End Function

    ''' <summary>
    ''' <code>Status =: Availed, Cracked, InActive, Matured, Partial</code>
    ''' </summary>
    ''' <returns>Status Of The Given Kitty</returns>
    Public Function Status() As String
        If IsAvailed Then
            If IsMatured() Then
                Return "Availed"
            Else
                Return "Cracked"
            End If
        ElseIf IsInactive Then
            Return "InActive"
        Else
            If IsMatured() Then
                Return "Matured"
            Else
                Return "Partial"
            End If
        End If
    End Function

#Region "Public Shared Function"
    Friend Shared Function GetListOfKittyTypes() As List(Of Integer)
        Dim temp_list As New List(Of Integer)
        Try
            Dim dr_kittyType As OleDbDataReader = DataReader("Select KittyType from CBox_Data Where (NOT (KittyType = ''))")
            While dr_kittyType.Read
                temp_list.Add(dr_kittyType("KittyType"))
            End While
            dr_kittyType.Close()
        Catch ex As Exception
            MessageBox.Show("Can't Get List Of KittyTypes [Kitty]: " + ex.Message)
        End Try
        Return temp_list
    End Function

    Friend Shared Function GetListOfKittyIntrests() As List(Of Integer)
        Dim temp_list As New List(Of Integer)
        Try
            Dim dr_kittyInterest As OleDbDataReader = DataReader("Select KittyIntrest from CBox_Data Where (NOT (KittyIntrest = ''))")
            While dr_kittyInterest.Read
                temp_list.Add(dr_kittyInterest("KittyIntrest"))
            End While
            dr_kittyInterest.Close()
        Catch ex As Exception
            MessageBox.Show("Can't Get List Of KittyInterests [Kitty]: " + ex.Message)
        End Try
        Return temp_list
    End Function

    Friend Shared Function GetListOfPaymentModes() As List(Of String)
        Dim temp_list As New List(Of String)
        Try
            Dim dr_Modes As OleDbDataReader = DataReader("Select PaymentModes from CBox_Data Where (NOT (PaymentModes = ''))")
            While dr_Modes.Read
                temp_list.Add(dr_Modes("PaymentModes"))
            End While
            dr_Modes.Close()
        Catch ex As Exception
            MessageBox.Show("Can't Get List Of PaymentModes [Kitty]: " + ex.Message)
        End Try
        Return temp_list
    End Function

    Friend Shared Function GetListOfAccounts() As List(Of String)
        Dim temp_list As New List(Of String)
        Try
            Dim dr_Accounts As OleDbDataReader = DataReader("Select Accounts from CBox_Data Where (NOT (Accounts = ''))")
            While dr_Accounts.Read
                temp_list.Add(dr_Accounts("Accounts"))
            End While
            dr_Accounts.Close()
        Catch ex As Exception
            MessageBox.Show("Can't Get List Of Accounts [Kitty]: " + ex.Message)
        End Try
        Return temp_list
    End Function

    Public Function Equals(other As Kitty) As Boolean Implements IEquatable(Of Kitty).Equals
        Return other IsNot Nothing AndAlso
               KittyUID = other.KittyUID
    End Function

#End Region

End Class

File: OrnaTrack/Kitty.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.OleDb;
using System.Linq;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace FirstApp
{

    public class Kitty
    {

        #region Properties
        public int KittyUID { get; private set; } = -1;

        public int KittyNo { get; set; } = -1;

        public int KittyType { get; set; } = -1;

        public int KittyInterest { get; set; } = -1;

        public int Duration { get; set; } = -1;

        public int CustomerID { get; set; } = -1;

        public string Notes { get; set; } = "";

        public bool IsAvailed { get; set; } = false;

        public bool IsInactive { get; set; } = false;

        public Dictionary<DateTime, int> Record { get; set; } = new Dictionary<DateTime, int>();
        public Dictionary<DateTime, List<Transaction>> Transactions { get; set; } = new Dictionary<DateTime, List<Transaction>>();

        private bool Initialized { get; set; } = false;

        public Customer Owner
        {
            get
            {
                try
                {
                    return new Customer(CustomerID);
                }
                catch (Exception ex)
                {
                    MessageBox.Show("Error [Kitty]/Owner This Kitty DoesNot Have A Associated Customer With It.", "Wrong Selection", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return null;
                }
            }
        }

        #endregion

        public Kitty(int UID = -1, bool InitializeKitty = false, bool completely = false)
        {
            if (UID == -1)
                return;
            KittyUID = UID;
            if (InitializeKitty)
                Initialize(completely);
        }

        public void Initialize(bool InitializeCompletely = true)
        {
            if (KittyUID == -1)
                return;
            if (Initialized)
                return;
            OleDbDataReader dr = (OleDbDataReader)ProjectModule.DataReader("Select * From Kitty_Data Where KittyUID=" + KittyUID);
            while (dr.Read())
            {
                KittyNo = Conversions.ToInteger(dr["KittyNo"]);
                CustomerID = Conversions.ToInteger(dr["CoustID"]);
                KittyType = Conversions.ToInteger(dr["KittyType"]);
                KittyInterest = Conversions.ToInteger(dr["Interest"]);
                Duration = Conversions.ToInteger(dr["Duration"]);
                IsAvailed = Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(dr["Availed"], "False", false)) ? false : true;
                IsInactive = Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(dr["InActive"], "False", false)) ? false : true;
                try
                {
                    Notes = dr["Notes"].ToString().Replace(ProjectModule.apostropheReplacer, "'");
                }
                catch
                {
                }
                if (InitializeCompletely)
                {
                    AddRecord(Conversions.ToString(dr["Dates"]));
                    Initialized = true;
                }
            }
            dr.Close();

        }

        public void Save()
        {
            if (KittyUID == -1)
            {
                if (CustomerID == -1)
                {
                    MessageBox.Show("[Kitty] Attempt To Save A Kitty Without A Customer To Own It. Can't Proceed Further Without A CustomerId.", "Illegal Selection", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }
                if (MessageBox.Show("[Kitty] Do You Want To Add New Kitty ?", "Confirm", MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.No)
                    return;
                ProjectModule.SqlCommand("INSERT INTO Kitty_Data (CoustID,KittyNo,KittyType,Dates,Interest,Duration,Availed,InActive,Notes)" + $"VALUES({CustomerID},{KittyNo},{KittyType},'{RecordString()}',{KittyInterest},{Duration},'{IsAvailed}','{IsInactive}','{Notes.Replace("'", ProjectModule.apostropheReplacer)}')");
            }
            else
            {
                if (CustomerID == -1)
                    return;
                if (MessageBox.Show("[Kitty] Do You Want To Update This Kitty ?", "Confirm", MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.No)
                    return;
                ProjectModule.SqlCommand($"UPDATE Kitty_Data set CoustID={CustomerID},KittyNo={KittyNo},KittyType={KittyType},Dates='{RecordString()}',Interest={KittyInterest},Duration={Duration}" + $",Availed ='{IsAvailed}',InActive='{IsInactive}',Notes='{Notes.Replace("'", ProjectModule.apostropheReplacer)}' WHERE KittyUID={KittyUID}");
            }
        }

        public void Dissolve()
        {
            try
            {
                if (MessageBox.Show("Are You Sure That You Want To Dissolve This Kitty ?", "Confirm", MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.No)
                    return;
                ProjectModule.SqlCommand($"DELETE FROM Kitty_Data WHERE KittyUID={KittyUID}");
            }
            catch (Exception ex)
            {
                MessageBox.Show($"Error [Kitty]/Dissolve: {ex.Message}");
            }
        }

        public void InitializeTransactions()
        {
            Transactions.Clear();
            OleDbDataReader dr = (OleDbDataReader)ProjectModule.DataReader($"Select TrnsID From Transaction_Data Where KittyIds LIKE '%,{KittyUID},%'");
            while (dr.Read())
            {
                var _trns = new Transaction(Conversions.ToInteger(dr[0]));
                if (!Transactions.ContainsKey(_trns.TrnsDate))
                {
                    Transactions.Add(_trns.TrnsDate, new List<Transaction>(new[] { _trns }));
                }
                else
                {
                    Transactions[_trns.TrnsDate].Add(_trns);
                }
            }
            dr.Close();
        }

        public bool AddRecord(string _entry)
        {
            try
            {
                if (string.IsNullOrEmpty(_entry))
                    return false;
                foreach (string i in _entry.Split(','))
                {
                    string _entryDate = i.Substring(0, 10).Trim();
                    int _entryAmount = Conversions.ToInteger(i.Substring(10));
                    if (Record.ContainsKey(Conversions.ToDate(_entryDate)))
                    {
                        if (Record[Conversions.ToDate(_entryDate)] + _entryAmount < KittyType)
                        {
                            Record.Remove(Conversions.ToDate(_entryDate));
                        }
                        else
                        {
                            Record[Conversions.ToDate(_entryDate)] += _entryAmount;
                        }
                    }
                    else if (_entryAmount >= KittyType)
                        Record.Add(Conversions.ToDate(_entryDate), _entryAmount);
                }

                var sorted = from pair in Record
                             orderby pair.Key
                             select pair;

                Record = sorted.ToDictionary(k => k.Key, v => v.Value);

                foreach (var i in Transactions.Where(f => !Record.ContainsKey(f.Key)))
                {

                }

                return true;
            }
            catch (Exception ex)
            {
                MessageBox.Show("Can't Add Entry To Record [Kitty]: " + ex.Message);
                return false;
            }
        }

        public string RecordString(string Seperator = ",")
        {
            if (Record is null)
                return null;
            return string.Join(Seperator, Record.Select(f => $"{f.Key.ToShortDateString()}{f.Value}").ToArray());
        }

        public int TotalAmount()
        {
            return KittyType * Duration;
        }

        public int GivenAmount()
        {
            return Record.Values.Sum();
        }

        public int GetInstalmentsCompleted()
        {
            return (int)Math.Round(Conversion.Int(GivenAmount() / (double)KittyType));
        }

        public int GetInstalmentsPending()  // Instalments Pending Till Date
        {
            if (Record.Count == 0)
                return default;
            var FirstDate = Record.First().Key;

            var ExpectedDate = FirstDate;
            int ExpectedInstalments = 0;

            while (ExpectedDate < DateTime.Today)  // Adding One To Get Instalments Expected Until Today (It Will Add One Extra Month At Last)
            {
                ExpectedDate = ExpectedDate.AddMonths(1);
                ExpectedInstalments += 1;
            }

            ExpectedInstalments -= 1; // Subtracting One To Get Instalments Expected Until Today

            ExpectedInstalments -= GetInstalmentsCompleted() - 1; // Subtracting Instalments Given And One More Instalment That Is The First Instalment

            if (ExpectedInstalments > GetInstalments_LeftForMaturity()) // If Instalments Due Past The Date Of Maturity
            {
                ExpectedInstalments = GetInstalments_LeftForMaturity();
            }

            if (GetInstalments_LeftForMaturity() == 0)
                return -1; // If Matured -> Return -1

            if (ExpectedInstalments < 0) // If SomeBody Paid More Than Necessary
            {
                return 0;
            }
            else
            {
                return ExpectedInstalments;
            }
        }

        public int GetInstalments_LeftForMaturity() // Instalments Pending For Maturity
        {
            return Conversion.Int(Duration - GetInstalmentsCompleted());
        }

        public bool IsActive()
        {
            Initialize(true);
            return !(IsMatured() | IsAvailed | IsInactive);
        }

        public bool IsMatured()
        {
            if (GivenAmount() >= KittyType * Duration)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        /// <summary>
    /// <code>Status =: Availed, Cracked, InActive, Matured, Partial</code>
    /// </summary>
    /// <returns>Status Of The Given Kitty</returns>
        public string Status()
        {
            if (IsAvailed)
            {
                if (IsMatured())
                {
                    return "Availed";
                }
                else
                {
                    return "Cracked";
                }
            }
            else if (IsInactive)
            {
                return "InActive";
            }
            else if (IsMatured())
            {
                return "Matured";
            }
            else
            {
                return "Partial";
            }
        }

        #region Public Shared Function
        internal static List<int> GetListOfKittyTypes()
        {
            var temp_list = new List<int>();
            try
            {
                OleDbDataReader dr_kittyType = (OleDbDataReader)ProjectModule.DataReader("Select KittyType from CBox_Data Where (NOT (KittyType = ''))");
                while (dr_kittyType.Read())
                    temp_list.Add(Conversions.ToInteger(dr_kittyType["KittyType"]));
                dr_kittyType.Close();
            }
            catch (Exception ex)
            {
                MessageBox.Show("Can't Get List Of KittyTypes [Kitty]: " + ex.Message);
            }
            return temp_list;
        }

        internal static List<int> GetListOfKittyIntrests()
        {
            var temp_list = new List<int>();
            try
            {
                OleDbDataReader dr_kittyInterest = (OleDbDataReader)ProjectModule.DataReader("Select KittyIntrest from CBox_Data Where (NOT (KittyIntrest = ''))");
                while (dr_kittyInterest.Read())
                    temp_list.Add(Conversions.ToInteger(dr_kittyInterest["KittyIntrest"]));
                dr_kittyInterest.Close();
            }
            catch (Exception ex)
            {
                MessageBox.Show("Can't Get List Of KittyInterests [Kitty]: " + ex.Message);
            }
            return temp_list;
        }

        internal static List<string> GetListOfPaymentModes()
        {
            var temp_list = new List<string>();
            try
            {
                OleDbDataReader dr_Modes = (OleDbDataReader)ProjectModule.DataReader("Select PaymentModes from CBox_Data Where (NOT (PaymentModes = ''))");
                while (dr_Modes.Read())
                    temp_list.Add(Conversions.ToString(dr_Modes["PaymentModes"]));
                dr_Modes.Close();
            }
            catch (Exception ex)
            {
                MessageBox.Show("Can't Get List Of PaymentModes [Kitty]: " + ex.Message);
            }
            return temp_list;
        }

        internal static List<string> GetListOfAccounts()
        {
            var temp_list = new List<string>();
            try
            {
                OleDbDataReader dr_Accounts = (OleDbDataReader)ProjectModule.DataReader("Select Accounts from CBox_Data Where (NOT (Accounts = ''))");
                while (dr_Accounts.Read())
                    temp_list.Add(Conversions.ToString(dr_Accounts["Accounts"]));
                dr_Accounts.Close();
            }
            catch (Exception ex)
            {
                MessageBox.Show("Can't Get List Of Accounts [Kitty]: " + ex.Message);
            }
            return temp_list;
        }

        #endregion

    }

    public class Transaction : ICloneable
    {
        public int TrnsID { get; set; } = -1;
        public DateTime TrnsDate { get; set; }
        public string Mode { get; set; } = "";
        public string Account { get; set; }
        public int Amount { get; set; } = -1;
        public string RefNo { get; set; }
        public string Notes { get; set; }
        public Dictionary<int, int> KittyIds { get; set; } = new Dictionary<int, int>();
        public List<Kitty> InitialisedKitty
        {
            get
            {
                return (from i in KittyIds
                        select new Kitty(i.Key, true)).ToList();
            }
        }

        public Transaction(int _id)
        {
            if (_id == -1)
                return;
            OleDbDataReader dr = (OleDbDataReader)ProjectModule.DataReader($"Select * From Transaction_Data Where TrnsID={_id}");
            while (dr.Read())
            {
                TrnsDate = DateTime.Parse(Conversions.ToString(dr["TrnsDate"]));
                Mode = Conversions.ToString(dr["Mode"]);
                Account = Conversions.ToString(dr["Account"]);
                Amount = Conversions.ToInteger(dr["Amount"]);
                RefNo = Conversions.ToString(dr["RefNo"] is DBNull ? "" : dr["RefNo"]);

                if (!(dr["KittyIds"] is DBNull))
                {
                    foreach (var i in dr["KittyIds"].ToString().Split(','))
                    {
                        if (!(i.Length > 0))
                            continue;
                        KittyIds.Add(Conversions.ToInteger(Conversion.Int(i)), -1);
                    }
                    int _index = 0;
                    foreach (var i in dr["Structure"].ToString().Split(','))
                    {
                        if (!(i.Length > 0))
                            continue;
                        KittyIds[KittyIds.ElementAt(_index).Key] = Conversions.ToInteger(Conversion.Int(i));
                        _index += 1;
                    }
                }

                Notes = Conversions.ToString(dr["Notes"] is DBNull ? "" : dr["Notes"]);
            }
            dr.Close();
            TrnsID = _id;
        }

        public int Save()
        {
            if (TrnsID < 0)
            {
                bool res = ProjectModule.SqlCommand("Insert Into Transaction_Data (TrnsDate,Mode,Account,Amount,RefNo,KittyIds,Notes,Structure) " + $"Values ('{TrnsDate.ToShortDateString()}','{Mode}','{Account}','{Amount}','{RefNo}',',{string.Join(",", KittyIds.Keys)},','{Notes}','{string.Join(",", KittyIds.Values)}')");
                if (res)
                {
                    OleDbDataReader dr = (OleDbDataReader)ProjectModule.DataReader("Select Max(TrnsID) From Transaction_Data");
                    while (dr.Read())
                    {
                        if (!(dr[0] is DBNull))
                        {
                            TrnsID = Conversions.ToInteger(Conversion.Int(dr[0]));
                            return Conversions.ToInteger(Conversion.Int(dr[0]));
                        }
                    }
                }
            }
            else
            {
                ProjectModule.SqlCommand("Update Transaction_Data Set " + $"TrnsDate='{TrnsDate.ToShortDateString()}',Mode='{Mode}',Account='{Account}',Amount='{Amount}',RefNo='{RefNo}'" + $",KittyIds=',{string.Join(",", KittyIds.Keys)},',Notes='{Notes}',Structure='{string.Join(",", KittyIds.Values)}' Where TrnsID={TrnsID}");
            }
            return -1;
        }

        public void Delete(int _kittyID = -1)
        {
            if (_kittyID != -1)
            {
                if (KittyIds.Count > 1)
                {
                    KittyIds.Remove(_kittyID);
                }
            }
            ProjectModule.SqlCommand($"Delete From Transaction_Data Where TrnsID={TrnsID}");
        }

        public override int GetHashCode()
        {
            return (TrnsID, TrnsDate, Mode, Account, Amount, RefNo, KittyIds, Notes).GetHashCode();
        }

        public object Clone()
        {
            var x = MemberwiseClone();
            ((dynamic)x).KittyIds = KittyIds.ToDictionary(g => g.Key, g => g.Value);
            return x;
        }
    }
}